beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockOperationSelectorFactory.java

```diff
@@ -172,174 +166,197 @@ public Consumer<BeaconBlockBodyBuilder> createSelector(
       }
 
       // Execution Payload / Execution Payload Header / KZG Commitments
+      final SafeFuture<Void> completionFuture;
       if (bodyBuilder.supportsExecutionPayload()) {
         final SchemaDefinitions schemaDefinitions =
             spec.atSlot(blockSlotState.getSlot()).getSchemaDefinitions();
-        setExecutionData(
-            bodyBuilder,
-            blinded,
-            schemaDefinitions,
-            parentRoot,
-            blockSlotState,
-            blockProductionPerformance);
+
+        completionFuture =
+            forkChoiceNotifier
+                .getPayloadId(parentRoot, blockSlotState.getSlot())
+                .thenCompose(
+                    executionPayloadContext ->
+                        setExecutionData(
+                            executionPayloadContext,
+                            bodyBuilder,
+                            requestedBlinded,
+                            schemaDefinitions,
+                            blockSlotState,
+                            blockProductionPerformance));
+      } else {
+        completionFuture = SafeFuture.COMPLETE;
       }
 
       blockProductionPerformance.beaconBlockPrepared();
+
+      return completionFuture;
     };
   }
 
-  private void setExecutionData(
+  private SafeFuture<Void> setExecutionData(
+      final Optional<ExecutionPayloadContext> executionPayloadContext,
       final BeaconBlockBodyBuilder bodyBuilder,
-      final boolean blinded,
+      final Optional<Boolean> requestedBlinded,
       final SchemaDefinitions schemaDefinitions,
-      final Bytes32 parentRoot,
       final BeaconState blockSlotState,
       final BlockProductionPerformance blockProductionPerformance) {
-    final SafeFuture<Optional<ExecutionPayloadContext>> executionPayloadContextFuture =
-        forkChoiceNotifier.getPayloadId(parentRoot, blockSlotState.getSlot());
+
+    // if blinded flow is explicitly requested, we should try to use it
+    // otherwise, we should use it only if we have a validator registration
+    final boolean shouldTryBuilderFlow =
+        requestedBlinded.orElse(true)
+            || executionPayloadContext
+                .map(ExecutionPayloadContext::isValidatorRegistrationPresent)
+                .orElse(false);
+
+    // we should try to return unblinded content only if we try the builder flow with no explicit
+    // request
+    final boolean setUnblindedContentIfPossible =
```

Comment:
how about `setLocalPayloadIfBuilderFailed`? Current name is too abstract, I think we'd better use name more tied with domain. Also comment could describe why and in what case this variable is used